### PR TITLE
🏗️ Support external auth provider login.

### PR DESCRIPTION
**Pull request by Danieloni1**

### CODE REVIEW

**Comment by Danieloni1** [comment]
@danpe I've organized the messy password login code, and added support for external auth provider (for GeoDome).
Still need to make the request to the external auth provider and continue with the existing logic in case I don't get 403, but I'll do that after I get the account in GeoDome and their API.

File: clientapi/routing/login.go [hunk 1]
```diff
@@ -42,26 +44,28 @@ type flow struct {
...
 ) util.JSONResponse {
 	if req.Method == http.MethodGet {
-		// TODO: support other forms of login other than password, depending on config options
+		loginFlows := []flow{{Type: authtypes.LoginTypePassword}}
+		if len(cfg.Derived.ApplicationServices) > 0 {
+			loginFlows = append(loginFlows, flow{Type: authtypes.LoginTypeApplicationService})
+			loginFlows = append(loginFlows, flow{Type: authtypes.LoginTypeDummy})
+			loginFlows = append(loginFlows, flow{Type: authtypes.LoginTypeEmail})
+			loginFlows = append(loginFlows, flow{Type: authtypes.LoginTypeJwt})
+			loginFlows = append(loginFlows, flow{Type: authtypes.LoginTypeGeodome})
```
Comment (danpe) [inline clientapi/routing/login.go:60]: @Danieloni1 why appending Geodome twice?

File: clientapi/auth/user_interactive.go [hunk 2]
```diff
@@ -72,6 +72,8 @@ type Login struct {
 	// Thus a pointer is needed to differentiate between the two
 	InitialDisplayName *string `json:"initial_device_display_name"`
 	DeviceID           *string `json:"device_id"`
+
+	ExternalAuthResp interface{} `json:"external_auth_response"`
```
Comment (danpe) [inline clientapi/auth/user_interactive.go:76]: @Danieloni1 do we need this as part of `UserInteractive`?

File: clientapi/routing/login.go [hunk 3]
```diff
@@ -28,10 +29,11 @@ import (
...
 type loginResponse struct {
-	UserID      string          `json:"user_id"`
-	AccessToken string          `json:"access_token"`
-	HomeServer  spec.ServerName `json:"home_server"`
-	DeviceID    string          `json:"device_id"`
+	UserID           string          `json:"user_id"`
+	AccessToken      string          `json:"access_token"`
+	HomeServer       spec.ServerName `json:"home_server"`
+	DeviceID         string          `json:"device_id"`
+	ExternalAuthResp interface{}     `json:"external_auth_response,omitempty"`
```
Comment (danpe) [inline clientapi/routing/login.go:36]: `external_response`

**Review by danpe — CHANGES_REQUESTED** [review]

**Pull request closed without merging**